cpp_exercise/src/lib/ipcshmlib.cc

```diff
@@ -0,0 +1,247 @@
+
+
+    
+
+    while(_shm_fd == -1){
+     std::cout<<"          Waiting for the Receiver... "<<std::endl;
+     _shm_fd = shm_open(_shm_name.c_str(), O_RDWR, 0666);
+     _sem_sender = sem_open(_semaphoresender_name.c_str(),0);
+     _sem_receiver = sem_open(_semaphorereceiver_name.c_str(),0);
+
+         sleep(5);
+    }
+    
+    _err = ftruncate(_shm_fd, _shm_size);
+    if(_err<0){
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCSender ERROR: Cannot truncate the Shared Memory.");
+
+    }
+ 
+   
+
+    _ptr = (struct _shm_data_struct *)mmap(0, _shm_size, PROT_WRITE, MAP_SHARED, _shm_fd, 0);
+
+    if(_ptr == (struct _shm_data_struct *)-1){
+        std::cout<<strerror(errno)<<std::endl;
+        throw IPCException("IPCSender ERROR: Cannot map the Shared Memory.");
+    }
+    
+    
+    FileHandler fd(_file_name);
+
+    fd.openForReading();
+
+    std::cout<<"          File Name: "<<_file_name<<std::endl;
+    std::cout<<"          File Size: "<<fd.getSize()<<" bytes"<<std::endl;
+
+    
+    std::cout<<"          SharedMemory is Open: Trying to send the file  "<<std::endl;
+  
+
+    while(true){
+
+                
+       _err = sem_wait(_sem_receiver);
+       if(_err<0){
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCSender ERROR: Semaphore receiver.");
+
+         }   
+
+        std::vector<unsigned char>  buffer; 
+        
+        buffer = fd.readFile();
+
+        
+        _shm_data_struct tempstruct;
+        tempstruct.datasize = buffer.size();
+
+        std::copy(buffer.begin(), buffer.end(),tempstruct.data);
+    
+         if(buffer.size()==0){
+               memcpy(_ptr,&tempstruct, sizeof(tempstruct));
+              _err = sem_post(_sem_sender);
+              if(_err<0){
+                  std::cout<<strerror(errno)<<std::endl;
+                 throw IPCException("IPCSender ERROR: Semaphore Sender.");
+
+                   }  
+               break;
+             }
+
+        memcpy(_ptr,&tempstruct, sizeof(tempstruct));
+        _err =  sem_post(_sem_sender);
+        if(_err<0){
+                  std::cout<<strerror(errno)<<std::endl;
+                 throw IPCException("IPCSender ERROR: Semaphore Sender.");
+
+        } 
+
+    }
+   
+   
+       std::cout<<"          Successfully written to the SharedMemory "<<std::endl;
+       std::cout<<"          SharedMemory is Closed  "<<std::endl;
+
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"****************************IPCSender: SharedMemory PROTOCOL**************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+    
+
+}
+SharedMemorySender::~SharedMemorySender(){
```

Comment:
If a constructor of a class throws, the destructor of this class won't be called. So the sharedmemory won't be munmap/unlink.